### PR TITLE
Copilot chat: Avoid duplication in document memories

### PR DESCRIPTION
**Pull request by TaoChenOSU**
### Motivation and Context

Importing documents in Copilot chat will not check for duplication. This will potentially create issues where the same documents are uploaded multiple times.

### Description

1. Enable the document import controller to check for duplication before saving a new memory to the document memory.
2. Allow developers to configure the threshold of duplication (similarity score) in the settings.
3. Move a couple hard-coded similarity score settings to settings.

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs [hunk 1]
```diff
@@ -46,4 +46,10 @@ public class DocumentMemoryOptions
     /// </summary>
     [Range(0, int.MaxValue)]
     public int FileSizeLimit { get; set; } = 1000000;
+
+    /// <summary>
+    /// Similarity threshold to avoid document memory duplication when importing documents to memory.
+    /// </summary>
+    [Range(0.0, 1.0)]
+    public double DeduplicationSimilarityThreshold { get; set; } = 0.8;
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:54]: Could you add a comment on the logic of the default value? I naively feel its low for dedupe.

File: samples/apps/copilot-chat-app/webapi/appsettings.json [hunk 2]
```diff
@@ -129,13 +129,16 @@
...
+  //   another document that is already in memory.
   //
   "DocumentMemory": {
     "GlobalDocumentCollectionName": "global-documents",
     "UserDocumentCollectionNamePrefix": "user-documents-",
     "DocumentLineSplitMaxTokens": 30,
     "DocumentParagraphSplitMaxLines": 100,
-    "FileSizeLimit": 1000000
+    "FileSizeLimit": 1000000,
+    "DeduplicationSimilarityThreshold": 0.8
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/appsettings.json:141]: @glahaye  - do we need to run all appsettings into the deployment logic, or do default  values work as well?

**Review by adrianwyatt — CHANGES_REQUESTED** [review]

File: samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs (hunk 1 above)
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:56]: It's actually the opposite. The lower the value, the stronger the deduplication will be.
I have added a comments in both the source and the setting Json.

File: samples/apps/copilot-chat-app/webapi/appsettings.json (hunk 2 above)
Comment (glahaye) [reply samples/apps/copilot-chat-app/webapi/appsettings.json:143]: Right now, there is nothing related to DocumentMemory in the deployment templates.  So, all the config we get is from the appsettings.json document.

These values can be overridden online through the Azure Portal too.

Comment (glahaye) [reply samples/apps/copilot-chat-app/webapi/appsettings.json:143]: So, nothing to change in templates :)

File: samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs [hunk 3]
```diff
@@ -46,4 +46,12 @@ public class DocumentMemoryOptions
     /// </summary>
     [Range(0, int.MaxValue)]
     public int FileSizeLimit { get; set; } = 1000000;
+
+    /// <summary>
+    /// Similarity threshold to avoid document memory duplication when importing documents to memory.
+    /// The higher the value, the more similar document memories will be allowed.
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:52]: That comment makes my head spin.

Is it:
The higher the value, the more documents have to be similar to be eliminated as duplicates
OR
The higher the value, the more documents can be similar without being eliminated as duplicates

File: samples/apps/copilot-chat-app/webapi/Controllers/DocumentImportController.cs [hunk 4]
```diff
@@ -181,11 +181,21 @@ private async Task ParseDocumentContentToMemoryAsync(IKernel kernel, string cont
...
                 collection: targetCollectionName,
-                text: paragraph,
-                id: Guid.NewGuid().ToString(),
-                description: $"Document: {documentName}");
+                query: paragraph,
+                limit: 1,
+                minRelevanceScore: this._options.DeduplicationSimilarityThreshold
+            ).ToEnumerable();
+
+            if (!memories.Any())
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/webapi/Controllers/DocumentImportController.cs:191]: Don't you need to away before using memories?

Is SearchAsync returning an IAsyncEnumerable?

File: samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs (hunk 3 above)
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:52]: The second one

Comment (glahaye) [reply samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:52]: Yeah, maybe replace the two lines of the comment above (and in the settings) with the second line above.

File: samples/apps/copilot-chat-app/webapi/Controllers/DocumentImportController.cs (hunk 4 above)
Comment (TaoChenOSU) [reply samples/apps/copilot-chat-app/webapi/Controllers/DocumentImportController.cs:191]: `ToEnumerable` is applied on the IAsyncEnumerable instance.

File: samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs (hunk 3 above)
Comment (glahaye) [reply samples/apps/copilot-chat-app/webapi/Config/DocumentMemoryOptions.cs:52]: To me, this single line "The higher the value, the more documents can be similar without being eliminated as duplicates" is way easier to understand than the 2 that are currently there.  Unless you have a reason not to, just use that one line

**Review by adrianwyatt — DISMISSED** [review]

**Review by glahaye — APPROVED** [review]

**Pull request merged**